engine/access/rpc/backend/backend_transactions.go

```diff
@@ -1015,3 +1017,269 @@ func (b *backendTransactions) tryGetTransactionResultByIndex(
+
+// lookupTransactionErrorMessage returns transaction error message for specified transaction.
+// If an error message for transaction can be found in the cache then it will be used to serve the request, otherwise
+// an RPC call will be made to the EN to fetch that error message, fetched value will be cached in the LRU cache.
+func (b *backendTransactions) lookupTransactionErrorMessage(
+	ctx context.Context,
+	blockID flow.Identifier,
+	transactionID flow.Identifier,
+) (string, error) {
+	cacheKey := flow.MakeIDFromFingerPrint(append(blockID[:], transactionID[:]...))
+	value, cached := b.txErrorMessagesCache.Get(cacheKey)
+	if cached {
+		return value, nil
+	}
+
+	txResult, err := b.results.ByBlockIDTransactionID(blockID, transactionID)
+	if err != nil {
+		return "", rpc.ConvertStorageError(err)
+	}
+
+	// nothing to do tx has not failed
+	if !txResult.Failed {
+		return "", nil
+	}
+
+	execNodes, err := executionNodesForBlockID(ctx, blockID, b.executionReceipts, b.state, b.log)
+	if err != nil {
+		if IsInsufficientExecutionReceipts(err) {
+			return "", status.Errorf(codes.NotFound, err.Error())
+		}
+		return "", rpc.ConvertError(err, "failed to retrieve error message from any execution node", codes.Internal)
+	}
+	req := &execproto.GetTransactionErrorMessageRequest{
+		BlockId:       convert.IdentifierToMessage(blockID),
+		TransactionId: convert.IdentifierToMessage(transactionID),
+	}
+
+	resp, err := b.getTransactionErrorMessageFromAnyEN(ctx, execNodes, req)
+	if err != nil {
+		return "", fmt.Errorf("could not fetch error message from ENs: %w", err)
+	}
+	value = resp.ErrorMessage
+	b.txErrorMessagesCache.Add(cacheKey, value)
+	return value, nil
+}
+
+// lookupTransactionErrorMessageByIndex returns transaction error message for specified transaction using its index.
+// If an error message for transaction can be found in cache then it will be used to serve the request, otherwise
+// an RPC call will be made to the EN to fetch that error message, fetched value will be cached in the LRU cache.
+func (b *backendTransactions) lookupTransactionErrorMessageByIndex(
+	ctx context.Context,
+	blockID flow.Identifier,
+	index uint32,
+) (string, error) {
+	txResult, err := b.results.ByBlockIDTransactionIndex(blockID, index)
+	if err != nil {
+		return "", rpc.ConvertStorageError(err)
+	}
+
+	// nothing to do tx has not failed
+	if !txResult.Failed {
+		return "", nil
+	}
+
+	cacheKey := flow.MakeIDFromFingerPrint(append(blockID[:], txResult.TransactionID[:]...))
+	value, cached := b.txErrorMessagesCache.Get(cacheKey)
+	if cached {
+		return value, nil
+	}
+
+	execNodes, err := executionNodesForBlockID(ctx, blockID, b.executionReceipts, b.state, b.log)
+	if err != nil {
+		if IsInsufficientExecutionReceipts(err) {
+			return "", status.Errorf(codes.NotFound, err.Error())
+		}
+		return "", rpc.ConvertError(err, "failed to retrieve error message from any execution node", codes.Internal)
+	}
+	req := &execproto.GetTransactionErrorMessageRequest{
```

Comment:
this should use the `index` specific endpoint. the index methods are partly as a convenience when you're iterating all tx in a block, but also as a safety check to ensure you are getting exactly the tx you expect.
```suggestion
	req := &execproto.GetTransactionErrorMessageByIndexRequest{
```